loopy/preprocess.py

```diff
@@ -1996,6 +1996,246 @@ def realize_ilp(kernel):
+               if isinstance(arg, ArrayArg) and arg.shape[0] is not None]
+    gbf = VariableFinder(globals)
+    cf = ConditionalFinder()
+
+    new_insts = []
+    for inst in kernel.instructions:
+        if isinstance(inst, Assignment) and (inst.within_inames & set(cvec_inames)):
+            can_vectorize = True
+            should_simd = False
+
+            # reduction over globals are sequentialized
+            if can_vectorize and globals:
+                gbf.result = False
+                gbf(inst.assignee)
+                if gbf.result:
+                    can_vectorize = False
+
+            # some built-in functions cannot be vectorized
+            # (in general, ICC can vectorize some of them)
+            if can_vectorize:
+                function_finder.result = False
+                function_finder(inst.expression)
+                if function_finder.result:
+                    can_vectorize = False
+                    should_simd = True
+
+            # non privitized array indexed with c_vec iname cannot be vectorized
+            if can_vectorize:
+                for name in inst.dependency_names():
+                    if name in kernel.temporary_variables:
+                        ary = kernel.temporary_variables[name]
+                    elif name in kernel.arg_dict:
+                        ary = kernel.arg_dict[name]
+                        if not isinstance(ary, ArrayArg):
+                            continue
+                    else:
+                        continue
+
+                    if len(ary.dim_tags) < 1:
+                        continue
+
+                    if not isinstance(ary.dim_tags[-1], CVectorArrayDimTag):
+                        sf = SubscriptFinder(ary.name, set(cvec_inames))
+                        sf(inst.assignee)
+                        sf(inst.expression)
+                        if sf.result:
+                            can_vectorize = False
+                            break
+
+            # cannot have inames outside of subscirpt
+            if can_vectorize:
+                ovf = OutsideVariableFinder(
+                    list(inst.within_inames & set(cvec_inames)))
+                ovf(inst.expression)
+                if ovf.result:
+                    can_vectorize = False
+                    should_simd = True
+
+            # conditionals cannot be vectorized
+            if can_vectorize:
+                cf.result = False
+                cf(inst.expression)
+                if cf.result:
+                    can_vectorize = False
+                    should_simd = True
+
+            # constant rhs cannot be vectorized (GCC doesn't like it)
+            if can_vectorize:
+                avf = VariableFinder(list(inst.within_inames & set(cvec_inames)))
+                avf(inst.expression)
+                if not avf.result:
+                    import pymbolic.primitives as p
+                    rhs = inst.expression + p.Variable(name_zero_vec)
```

Comment:
The last failing test on firedrake side is due to this instruction. The p.Variable(name_zero_vec) is generated for one type, but potentially used for expressions with different types. Example: It might be defines as `double4 const zero_vec = { 0.0 };` and then used in an instruction` t1[i3] = t2[i3] + zero_vec;`, where t1 is int4. Then we get the error `error: assigning to 'int4' (vector of 4 'int' values) from incompatible type 'double4' (vector of 4 'double' values)`.

I guess the naive to resolve is to generate one zero_vec per type and then use the right one in this instruction dependent on the type of expression. The alternative might be a casting operation, but I do not know if that will be possible. Does anyone have another idea?

Reply:
Went for the first one, since it is computationally probably the cheaper one.